### ISSUE TITLE
Add "Save to Zotero (PDF)" to Zotero Connector submenu when right-clicking on a framed PDF

### ISSUE THREAD

**Issue by dstillman** [zotero/zotero-connectors#152]
This is useful when viewing a framed PDF that isn't recognized by a translator, since it avoids having to do This Frame -> Show Only This Frame in Firefox (and that's not even an option in Chrome?).

**Comment by dstillman** [comment]
This is `fx-parity` due to the removal of the open/save dialog modification. Previously, you could just click the download icon in the PDF viewer and save to Zotero that way.

**Comment by adomasven** [comment]
Do we have any example URLs?

**Comment by dstillman** [comment]
https://www.cs.tut.fi/~jkorpela/html/iframe-pdf.html

**Comment by adomasven** [comment]
Seems like we need to talk upstream if we want this. Chrome does not show addon context-menus in PDFs completely. Firefox does, but the finest way you can filter these are either by showing to all [frames](https://developer.mozilla.org/en-US/Add-ons/WebExtensions/API/contextMenus/ContextType) or matching [url regex](https://developer.mozilla.org/en-US/Add-ons/WebExtensions/API/contextMenus/create).

Although if we're taking this upstream we may as well just ask for #83

**Comment by dstillman** [comment]
<blockquote>Chrome does not show addon context-menus in PDFs completely</blockquote>
I have a 1Password menu item when right-clicking on a framed PDF in Chrome (but probably for all frames).

> Although if we're taking this upstream we may as well just ask for #83
I already did — they're very unlikely to do something tied to the current UI in that way. But there might be other options involving download notifications (which may or may not work for us).

Can we tell whether there's a PDF in an iframe from the parent page? What about adding a "Save to Zotero (PDF)" option to our own context menu in that case?

**Comment by adomasven** [comment]
> I have a 1Password menu item when right-clicking on a framed PDF in Chrome (but probably for all frames).

Oh yeah, that actually works.

> Can we tell whether there's a PDF in an iframe from the parent page? What about adding a "Save to Zotero (PDF)" option to our own context menu in that case?

We could check for frames with content-type `application/pdf` with web request intercepting. We could then add a "Save to Zotero (PDF)" with context set to frames, but that would add that context-menu option to all frames, including non-pdf ones. (Seems like we could actually do this properly in [safari](https://developer.apple.com/library/content/documentation/Tools/Conceptual/SafariExtensionGuide/AddingContextualMenuItems/AddingContextualMenuItems.html))

I'm not really a huge fan of adding the context option to all frames, but you'd have to, since there may be multiple frames with pdfs and you need to know which one to get.

**Comment by dstillman** [comment]
To be clear, I mean adding an option to the save button context-menu when there's a PDF iframe on the page, so that even if the main icon was for Web Page you'd still have PDF as an option. (I'm not worried about there being multiple. We could just grab the first one.)

**Comment by adomasven** [comment]
Alternative plan. Chrome `contextMenu.onClicked` event provides `frameId` and `frameUrl`, so we can just add *Save frame to Zotero as Snapshot* context-menu option for frames.

**Comment by dstillman** [comment]
> I mean adding an option to the save button context-menu when there's a PDF iframe on the page, so that even if the main icon was for Web Page you'd still have PDF as an option

Actually, thinking about this more, If there's an iframe with a PDF and a translator isn't triggering on the page, we should just show the PDF icon. That's almost always what people will want.

**Comment by adomasven** [comment]
> Actually, thinking about this more, If there's an iframe with a PDF and a translator isn't triggering on the page, we should just show the PDF icon. That's almost always what people will want.

The alternative plan is super elegant and easy to implement, so let's start with that one and if people are still very confused/annoyed we'll do this.

**Comment by dstillman** [comment]
> The alternative plan is super elegant

Right-clicking on a page and talking to users about "frames" is not remotely super elegant. The elegant solution here is make the action that people are going to want the vast majority of the time the default button click. That's the priority here.
